File: packages/core/src/estimate.ts
import { normalizeToPerToken } from './normalize.js';
import { add, clampNonNegative, exact, mul, order, scale, ZERO } from './range.js';
import { getModel, getRegion, resolveModel, selectBenchmark } from './resolve.js';
import type {
  BenchmarkRow,
  Estimate,
  EstimateFlag,
  EstimateOptions,
  Model,
  Range,
  Region,
  Source,
  TokenCounts,
  UsageEvent,
} from './types.js';

/**
 * Turns one exchange into energy, water and carbon.
 *
 * The one rule that governs every branch below: when we do not know something,
 * the result is null and a flag says why. Nothing here has a path that turns a
 * missing value into a zero and carries on.
 */

/** An estimate with no numbers, which is what an unknown model produces. */
const unknownEstimate = (
  event: UsageEvent,
  datasetVersion: string,
  regionCode: string,
  flags: EstimateFlag[],
  thinkingHandling: Estimate['basis']['thinkingHandling'],
  modelId: string | null = null,
): Estimate => ({
  eventId: event.id,
  modelId,
  datasetVersion,
  energyWh: null,
  waterMl: null,
  carbonG: null,
  basis: {
    benchmarkRowId: null,
    benchmarkParentId: null,
    methodology: null,
    boundary: null,
    regionCode,
    tokens: event.tokens,
    thinkingHandling,
    thinkingTokens: null,
    waterScope: 'on-site + off-site',
    carbonBasis: 'location-based',
    qualityScore: null,
    flags,
    sources: [],
  },
});

/**
 * Works out how many thinking tokens to charge.
 *
 * A reported count is used as it stands. An undisclosed count on a model that
 * can think becomes a range from zero to the model's high ratio, flagged, so the
 * interface can render it as a lower bound. A model that cannot think charges
 * nothing, and that is the only case where zero is the right answer.
 */
const thinkingTokensFor = (
  tokens: TokenCounts,
  model: Model,
  visibleOutput: number,
): { range: Range; handling: Estimate['basis']['thinkingHandling']; flag: EstimateFlag | null } => {
  if (typeof tokens.thinking === 'number') {
    return { range: exact(tokens.thinking), handling: 'reported', flag: null };
  }
  if (tokens.thinking === null && model.reasoning && model.thinkingRatio) {
    return {
      range: {
        low: 0,
        central: visibleOutput * model.thinkingRatio.central,
        high: visibleOutput * model.thinkingRatio.high,
      },
      handling: 'estimated',
      flag: 'thinking-unknown',
    };
  }
  if (tokens.thinking === null && model.reasoning) {
    // The model can think, we have no ratio, and we refuse to invent one.
    return { range: exact(0), handling: 'unknown-model', flag: 'thinking-unknown' };
  }
  return { range: ZERO, handling: 'not-applicable', flag: null };
};

const collectSources = (row: BenchmarkRow, parent: BenchmarkRow | null, region: Region): Source[] => {
  const sources: Source[] = [row.source];
  if (parent && parent.source.url !== row.source.url) sources.push(parent.source);
  sources.push(region.source);
  if (region.waterSource && region.waterSource.url !== region.source.url) {
    sources.push(region.waterSource);
  }
  return sources;
};

export const estimate = (event: UsageEvent, options: EstimateOptions): Estimate => {
  const { dataset } = options;
  const calibration = dataset.calibration;
  const waterScope = options.waterScope ?? 'on-site + off-site';
  const carbonBasis = options.carbonBasis ?? 'location-based';
  const at = options.at ?? event.timestamp.slice(0, 10);

  const flags: EstimateFlag[] = [];

  const requestedRegion = options.regionCode ?? event.regionHint;
  let region = getRegion(dataset, requestedRegion);
  if (!region) {
    region = getRegion(dataset, dataset.defaultRegion);
    flags.push('region-default');
  }
  if (!region) throw new Error(`Dataset has no default region "${dataset.defaultRegion}"`);
  const regionCode = region.code;

  const model = getModel(event.modelId, dataset) ?? resolveModel(event.modelRaw, dataset);
  if (!model) {
    return unknownEstimate(event, dataset.version, regionCode, [...flags, 'model-unknown'], 'unknown-model');
  }

  const row = selectBenchmark(model, dataset, { at, hosting: event.hosting });
  if (!row) {
    return unknownEstimate(
      event,
      dataset.version,
      regionCode,
      [...flags, 'no-benchmark'],
      'unknown-model',
      model.id,
    );
  }
  if (event.hosting === 'local' && row.pue.high > 1) flags.push('local-row-missing');

  const parent = row.proxyOf
    ? (dataset.benchmarks.find((candidate) => candidate.id === row.proxyOf) ?? null)
    : null;

  if (event.tokens.estimated) flags.push('tokens-estimated');

  // ------------------------------------------------------------------ tokens

  const visibleOutput = (event.tokens.output ?? 0) + (event.tokens.tool ?? 0);
  const thinking = thinkingTokensFor(event.tokens, model, event.tokens.output ?? 0);
  if (thinking.flag) flags.push(thinking.flag);
  const outputTokens = order(add(exact(visibleOutput), thinking.range));

  const hidden = calibration.hiddenContextFactor[event.surface] ?? exact(1);
  if (hidden.high > 1) flags.push('input-partial');
  if (event.tokens.input === undefined && event.tokens.output !== undefined) {
    if (!flags.includes('input-partial')) flags.push('input-partial');
  }

  const countedInput = event.tokens.input ?? 0;
  const cachedRead = event.tokens.cachedRead ?? 0;
  const cachedWrite = event.tokens.cachedWrite ?? 0;

  // A cache hit still moves the tokens through the model's attention, but nobody
  // publishes what it costs. The low bound assumes free, the high bound assumes half.
  const cacheShare = calibration.cachedReadShareOfInput;
  const writeShare = calibration.cacheWriteShareOfInput;
  const inputTokens = order(
    add(
      add(mul(exact(countedInput), hidden), scale(cacheShare, cachedRead)),
      scale(writeShare, cachedWrite),
    ),
  );

  // ------------------------------------------------------------------ energy

  const rates = normalizeToPerToken(row, { dataset, model, calibration });

  let energyWh: Range | null = null;
  if (rates) {
    for (const flag of rates.flags) if (!flags.includes(flag)) flags.push(flag);
    const raw = add(add(rates.interceptWh, mul(rates.inputWh, inputTokens)), mul(rates.outputWh, outputTokens));
    const withOverhead = row.energyIncludesPue ? raw : mul(raw, row.pue);
    energyWh = clampNonNegative(order(withOverhead));
  } else {
    flags.push('no-energy-figure');
  }

  const energyKwh = energyWh ? scale(energyWh, 1 / 1000) : null;

  // ------------------------------------------------------------------- scale

  /**
   * How much bigger this exchange is than the one the row describes. Only used
   * for rows that publish water or carbon directly and so cannot be scaled by
   * energy.
   */
  const directScale = (): Range => {
    const perPrompt = row.perPrompt;
    if (!perPrompt) return exact(1);
    const k = calibration.outputToInputEnergyWeight.central;
    const referenceWork = perPrompt.referenceInputTokens.central + k * perPrompt.referenceOutputTokens.central;
    if (referenceWork <= 0) return exact(1);
    const work = order({
      low: inputTokens.low + k * outputTokens.low,
      central: inputTokens.central + k * outputTokens.central,
      high: inputTokens.high + k * outputTokens.high,
    });
    return scale(work, 1 / referenceWork);
  };

  // ------------------------------------------------------------------- water

  let waterMl: Range | null = null;
  const directWater = row.perPrompt?.directWaterMl;
  if (directWater) {
    waterMl = clampNonNegative(order(mul(directWater, directScale())));
    if (!flags.includes('scaled-direct-figure')) flags.push('scaled-direct-figure');
  } else if (energyKwh) {
    // Litres per kilowatt hour, times kilowatt hours, times a thousand for millilitres.
    const onsite = row.waterOnsiteLPerKwh ?? ZERO;
    let litres = mul(energyKwh, onsite);
    if (waterScope !== 'on-site') {
      litres = add(litres, mul(energyKwh, region.waterOffsiteLPerKwh));
    }
    waterMl = clampNonNegative(order(scale(litres, 1000)));
    if (waterScope === 'lifecycle' && row.embodiedShareOfTotal && options.includeEmbodied !== false) {
      const uplift: Range = {
        low: 1 / Math.max(1e-9, 1 - row.embodiedShareOfTotal.low),
        central: 1 / Math.max(1e-9, 1 - row.embodiedShareOfTotal.central),
        high: 1 / Math.max(1e-9, 1 - row.embodiedShareOfTotal.high),
      };
      waterMl = order(mul(waterMl, uplift));
    }
  }

  // ------------------------------------------------------------------ carbon

  let carbonG: Range | null = null;
  const directCarbon = row.perPrompt?.directCarbonG;
  if (directCarbon) {
    carbonG = clampNonNegative(order(mul(directCarbon, directScale())));
    if (!flags.includes('scaled-direct-figure')) flags.push('scaled-direct-figure');
  } else if (energyKwh) {
    // Location-based by default. A provider's market-based factor is a claim about
    // what it bought, not about the electrons it burned, and it flatters the number
    // by three or four times, so it is opt in.
    const intensity =
      carbonBasis === 'provider-reported' && row.carbonGPerKwh ? row.carbonGPerKwh : region.gridGco2PerKwh;
    carbonG = clampNonNegative(order(mul(energyKwh, intensity)));
    if (waterScope === 'lifecycle' && row.embodiedShareOfTotal && options.includeEmbodied !== false) {
      const uplift: Range = {
        low: 1 / Math.max(1e-9, 1 - row.embodiedShareOfTotal.low),
        central: 1 / Math.max(1e-9, 1 - row.embodiedShareOfTotal.central),
        high: 1 / Math.max(1e-9, 1 - row.embodiedShareOfTotal.high),
      };
      carbonG = order(mul(carbonG, uplift));
    }
  }

  return {
    eventId: event.id,
    modelId: model.id,
    datasetVersion: dataset.version,
    energyWh,
    waterMl,
    carbonG,
    basis: {
      benchmarkRowId: row.id,
      benchmarkParentId: parent?.id ?? null,
      methodology: row.methodology,
      boundary: row.boundary,
      regionCode,
      tokens: event.tokens,
      thinkingHandling: thinking.handling,
      thinkingTokens: thinking.handling === 'not-applicable' ? null : thinking.range,
      waterScope,
      carbonBasis,
      qualityScore: row.qualityScore,
      flags,
      sources: collectSources(row, parent, region),
    },
  };
};

File: apps/cli-py/src/betteruseofai/estimate.py
"""Turning one exchange into energy, water and carbon. Mirrors packages/core/src/estimate.ts."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .engine import (
    ZERO,
    Range,
    add,
    clamp_non_negative,
    exact,
    get_region,
    mul,
    normalize_to_per_token,
    order,
    resolve_model,
    scale,
    select_benchmark,
    to_range,
)

WATER_SCOPES = ("on-site", "on-site + off-site", "lifecycle")


@dataclass
class TokenCounts:
    input: int | None = 0
    output: int | None = 0
    cached_read: int = 0
    cached_write: int = 0
    # A number is a reported count, None means the model can think and did not
    # say, and "absent" means the model cannot think at all. Collapsing any of
    # these into another is the exact failure this project exists to avoid.
    thinking: int | None | str = "absent"
    tool: int = 0
    estimated: bool = False
    estimator: str = "provider"


@dataclass
class UsageEvent:
    id: str
    surface: str
    hosting: str
    model_raw: str
    tokens: TokenCounts
    timestamp: str
    session_id: str | None = None
    region_hint: str | None = None
    meta: dict[str, Any] = field(default_factory=dict)


@dataclass
class Estimate:
    event_id: str
    model_id: str | None
    dataset_version: str
    energy_wh: Range | None
    water_ml: Range | None
    carbon_g: Range | None
    benchmark_row_id: str | None
    benchmark_parent_id: str | None
    methodology: str | None
    boundary: str | None
    region_code: str
    thinking_handling: str
    thinking_tokens: Range | None
    water_scope: str
    carbon_basis: str
    quality_score: int | None
    flags: list[str]
    sources: list[dict[str, Any]]


def _unknown(
    event: UsageEvent,
    dataset_version: str,
    region_code: str,
    flags: list[str],
    thinking_handling: str,
    model_id: str | None = None,
) -> Estimate:
    return Estimate(
        event_id=event.id,
        model_id=model_id,
        dataset_version=dataset_version,
        energy_wh=None,
        water_ml=None,
        carbon_g=None,
        benchmark_row_id=None,
        benchmark_parent_id=None,
        methodology=None,
        boundary=None,
        region_code=region_code,
        thinking_handling=thinking_handling,
        thinking_tokens=None,
        water_scope="on-site + off-site",
        carbon_basis="location-based",
        quality_score=None,
        flags=flags,
        sources=[],
    )


def _thinking_for(
    tokens: TokenCounts, model: dict[str, Any], visible_output: int
) -> tuple[Range, str, str | None]:
    """How many thinking tokens to charge, and whether we are guessing."""
    if isinstance(tokens.thinking, int) and not isinstance(tokens.thinking, bool):
        return exact(float(tokens.thinking)), "reported", None
    if tokens.thinking is None and model.get("reasoning"):
        ratio = to_range(model.get("thinkingRatio"))
        if ratio is not None:
            return (
                Range(0.0, visible_output * ratio.central, visible_output * ratio.high),
                "estimated",
                "thinking-unknown",
            )
        # The model can think, we have no ratio, and we refuse to invent one.
        return exact(0.0), "unknown-model", "thinking-unknown"
    return ZERO, "not-applicable", None


def _collect_sources(
    row: dict[str, Any], parent: dict[str, Any] | None, region: dict[str, Any]
) -> list[dict[str, Any]]:
    sources = [row["source"]]
    if parent is not None and parent["source"]["url"] != row["source"]["url"]:
        sources.append(parent["source"])
    sources.append(region["source"])
    water_source = region.get("waterSource")
    if water_source and water_source["url"] != region["source"]["url"]:
        sources.append(water_source)
    return sources


def estimate(
    event: UsageEvent,
    dataset: dict[str, Any],
    region_code: str | None = None,
    water_scope: str = "on-site + off-site",
    carbon_basis: str = "location-based",
    include_embodied: bool = True,
    at: str | None = None,
) -> Estimate:
    calibration = dataset["calibration"]
    at = at or event.timestamp[:10]
    flags: list[str] = []

    requested = region_code or event.region_hint
    region = get_region(dataset, requested)
    if region is None:
        region = get_region(dataset, dataset["defaultRegion"])
        flags.append("region-default")
    if region is None:
        raise RuntimeError(f"Dataset has no default region \"{dataset['defaultRegion']}\"")
    resolved_region = region["code"]

    model = resolve_model(event.model_raw, dataset)
    if model is None:
        return _unknown(
            event, dataset["version"], resolved_region, [*flags, "model-unknown"], "unknown-model"
        )

    row = select_benchmark(model, dataset, at=at, hosting=event.hosting)
    if row is None:
        return _unknown(
            event,
            dataset["version"],
            resolved_region,
            [*flags, "no-benchmark"],
            "unknown-model",
            model["id"],
        )
    if event.hosting == "local" and row["pue"]["high"] > 1:
        flags.append("local-row-missing")

    parent = (
        next((r for r in dataset["benchmarks"] if r["id"] == row.get("proxyOf")), None)
        if row.get("proxyOf")
        else None
    )

    if event.tokens.estimated:
        flags.append("tokens-estimated")

    # ------------------------------------------------------------- tokens

    visible_output = (event.tokens.output or 0) + (event.tokens.tool or 0)
    thinking_range, thinking_handling, thinking_flag = _thinking_for(
        event.tokens, model, event.tokens.output or 0
    )
    if thinking_flag:
        flags.append(thinking_flag)
    output_tokens = order(add(exact(float(visible_output)), thinking_range))

    hidden = to_range(calibration["hiddenContextFactor"].get(event.surface)) or exact(1.0)
    if hidden.high > 1:
        flags.append("input-partial")
    if (
        event.tokens.input is None
        and event.tokens.output is not None
        and "input-partial" not in flags
    ):
        flags.append("input-partial")

    counted_input = float(event.tokens.input or 0)
    cached_read = float(event.tokens.cached_read or 0)
    cached_write = float(event.tokens.cached_write or 0)

    cache_share = to_range(calibration["cachedReadShareOfInput"])
    write_share = to_range(calibration["cacheWriteShareOfInput"])
    input_tokens = order(
        add(
            add(mul(exact(counted_input), hidden), scale(cache_share, cached_read)),
            scale(write_share, cached_write),
        )
    )

    # ------------------------------------------------------------- energy

    rates = normalize_to_per_token(row, dataset, model)
    energy_wh: Range | None = None
    if rates is not None:
        for flag in rates.flags:
            if flag not in flags:
                flags.append(flag)
        raw = add(
            add(rates.intercept_wh, mul(rates.input_wh, input_tokens)),
            mul(rates.output_wh, output_tokens),
        )
        with_overhead = raw if row["energyIncludesPue"] else mul(raw, to_range(row["pue"]))
        energy_wh = clamp_non_negative(order(with_overhead))
    else:
        flags.append("no-energy-figure")

    energy_kwh = scale(energy_wh, 1 / 1000) if energy_wh is not None else None

    def direct_scale() -> Range:
        per_prompt = row.get("perPrompt")
        if not per_prompt:
            return exact(1.0)
        k = float(calibration["outputToInputEnergyWeight"]["central"])
        reference = (
            float(per_prompt["referenceInputTokens"]["central"])
            + k * float(per_prompt["referenceOutputTokens"]["central"])
        )
        if reference <= 0:
            return exact(1.0)
        work = order(
            Range(
                input_tokens.low + k * output_tokens.low,
                input_tokens.central + k * output_tokens.central,
                input_tokens.high + k * output_tokens.high,
            )
        )
        return scale(work, 1 / reference)

    def embodied_uplift() -> Range | None:
        share = to_range(row.get("embodiedShareOfTotal"))
        if share is None or water_scope != "lifecycle" or not include_embodied:
            return None
        return Range(
            1 / max(1e-9, 1 - share.low),
            1 / max(1e-9, 1 - share.central),
            1 / max(1e-9, 1 - share.high),
        )

    # -------------------------------------------------------------- water

    water_ml: Range | None = None
    direct_water = to_range((row.get("perPrompt") or {}).get("directWaterMl"))
    if direct_water is not None:
        water_ml = clamp_non_negative(order(mul(direct_water, direct_scale())))
        if "scaled-direct-figure" not in flags:
            flags.append("scaled-direct-figure")
    elif energy_kwh is not None:
        onsite = to_range(row.get("waterOnsiteLPerKwh")) or ZERO
        litres = mul(energy_kwh, onsite)
        if water_scope != "on-site":
            litres = add(litres, mul(energy_kwh, to_range(region["waterOffsiteLPerKwh"])))
        water_ml = clamp_non_negative(order(scale(litres, 1000)))
        uplift = embodied_uplift()
        if uplift is not None:
            water_ml = order(mul(water_ml, uplift))

    # ------------------------------------------------------------- carbon

    carbon_g: Range | None = None
    direct_carbon = to_range((row.get("perPrompt") or {}).get("directCarbonG"))
    if direct_carbon is not None:
        carbon_g = clamp_non_negative(order(mul(direct_carbon, direct_scale())))
        if "scaled-direct-figure" not in flags:
            flags.append("scaled-direct-figure")
    elif energy_kwh is not None:
        # Location-based by default. A market-based factor is a claim about what
        # the provider bought, not about the electrons it burned.
        row_intensity = to_range(row.get("carbonGPerKwh"))
        intensity = (
            row_intensity
            if carbon_basis == "provider-reported" and row_intensity is not None
            else to_range(region["gridGco2PerKwh"])
        )
        carbon_g = clamp_non_negative(order(mul(energy_kwh, intensity)))
        uplift = embodied_uplift()
        if uplift is not None:
            carbon_g = order(mul(carbon_g, uplift))

    return Estimate(
        event_id=event.id,
        model_id=model["id"],
        dataset_version=dataset["version"],
        energy_wh=energy_wh,
        water_ml=water_ml,
        carbon_g=carbon_g,
        benchmark_row_id=row["id"],
        benchmark_parent_id=parent["id"] if parent else None,
        methodology=row["methodology"],
        boundary=row["boundary"],
        region_code=resolved_region,
        thinking_handling=thinking_handling,
        thinking_tokens=None if thinking_handling == "not-applicable" else thinking_range,
        water_scope=water_scope,
        carbon_basis=carbon_basis,
        quality_score=row["qualityScore"],
        flags=flags,
        sources=_collect_sources(row, parent, region),
    )
